### PR TITLE
Custom layer example

### PR DESCRIPTION
**Pull request by bileschi**
Illustrates how to define, train, and infer from a model with a custom layer in tfjs.    
image

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-examples/61)

### CODE REVIEW

**Comment by caisq** [comment]
Thanks for making this example! I think having an example showing people how to write a custom layer is very important, as we have seen in the questions asked by our users. However, as the PR stands, I think we can do a better job highlighting the critical part. IIUC, much of the code in this example are the same as in the MNIST example, only a small and somewhat hard-to-find part of the code has to do with the actual custom layer. The MNIST example code is a little long, because much of it has to do with data loading and visualization. I suggest either highlight the custom layer code in an effective way, or consider reducing the complexity and size of the non-critical code. The latter may involve sacrificing some elegance of the visualization and/or truthfulness of the training data, but that should be okay given the goal of this example.

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[custom-layer/index.js, line 57 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-examples/61#-LAFSprK2_Gd9bfE90gK:-LAFSprK2_Gd9bfE90gL:b-qgcbu4) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/00f339c232f250dfa6fa024651969e63b18bc7ab/custom-layer/index.js#L57)):*
> ```
> call(inputs, kwargs)
> ```

maybe add some brief description, in words or math equations, what antirectifying does

---

*[custom-layer/index.js, line 104 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-examples/61#-LAFT1LUAAgKvbjlzUDk:-LAFT1LUAAgKvbjlzUDl:b-b3lvf9) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/00f339c232f250dfa6fa024651969e63b18bc7ab/custom-layer/index.js#L104)):*
> ```
> new Antirectifier()
> ```

Add a comment to highlight this is the place the custom Layer is used. (Same below.)

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-examples/61#-:-LAFTB8i2qqGm_T3HkAK:bw9tnoy)*

**Comment by bileschi** [comment]
Thanks for your review.   I agree that the custom layer code isn't very salient, but I'd rather not sacrifice the correctness or visual impact.  Lets focus on drawing attention to the important bits, and separate out the display functionality.  What do you think about extracting the model code into separate .js files?  Perhaps one for 'stock_mnist_model' and one for 'custom_layer_mnist_model'?

---

Review status: 0 of 9 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[custom-layer/index.js, line 57 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-examples/61#-LAFSprK2_Gd9bfE90gK:-LAInscK9YdXHtOm4XXm:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/00f339c232f250dfa6fa024651969e63b18bc7ab/custom-layer/index.js#L57)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> call(inputs, kwargs)
> ```

maybe add some brief description, in words or math equations, what antirectifying does
</blockquote></details>

Done.

---

*[custom-layer/index.js, line 104 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-examples/61#-LAFT1LUAAgKvbjlzUDk:-LAImhP55QFSMogCFYHR:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/00f339c232f250dfa6fa024651969e63b18bc7ab/custom-layer/index.js#L104)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> new Antirectifier()
> ```

Add a comment to highlight this is the place the custom Layer is used. (Same below.)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-examples/61#-:-LAImFxh89a5O4wKQF8H:b-pb5m5r)*

**Comment by nsthorat** [comment]
Just curious, is there a reason we're using MNIST here? Is there something simpler so we don't explode the number of MNIST examples (these all need major refactoring once we remove this hacky one-off data solution). I'm curious if we can reduce the scope of this example.

---

Review status: 0 of 9 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-examples/61#-:-LAIulsv9xQDL0RkzpHA:bk8favy)*

**Comment by bileschi** [comment]
I used MNIST for two reasons.  First because it's what's used in the [analogous example in python Keras](https://github.com/keras-team/keras/blob/master/examples/antirectifier.py).  Second because it was available / convenient as the hello-world of ML problems.  I'm not wedded to it.  Do you have any suggestions for a simpler substitute?

---







*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-examples/61#-:-LAJGoUeCb34iKWYLhpb:bx3pc1e)*

**Comment by caisq** [comment]
It seems to me that the Antirectifier custom layer you are using here is quite general. I doesn't have to be used with MNIST or any image models. Does it make sense to use it to built a simpler model that operates on very simple synthetic data?

---

Review status: 0 of 9 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-examples/61#-:-LAJHqYm1oe_Kr61B5eW:b-h01wvw)*

**Comment by bileschi** [comment]
I understand the motivation to extract what's new, i.e., the custom layer, from what's stock.  MNIST is very well known, however.  I'm not sure that a toy example would be any simpler or easier for a reader to understand.

I could remove all the classification work, and show how the custom layer applies to an input image, with two output images.  This would clearly illustrate the effect of the layer.  It would, however, remove the part showing that custom layers can be trained.

Thoughts?

**Comment by caisq** [comment]
It's a good point that for the purpose  of showcasing how to write and use custom layers, you don't need the training. The logic for custom layer is orthogonal to the logic for training (for all cases except the most complicated ones in which you want to override gradient updates, but we are not concerned with that in this example). Just doing inference with a hard-coded example sounds good to me.

---

Review status: 0 of 9 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-examples/61#-:-LAJMpYi5nabLCixLgNd:bkiijr1)*

**Comment by bileschi** [comment]
Updated to show just the layer.apply on an input image.

image

**Comment by caisq** [comment]
I see you are still pushing changes, Stan. Whenever this PR is ready for another look, please let me know. Thanks.

---

Review status: 0 of 9 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-examples/61#-:-LAJNCtAEzgy6nNiawi9:b98ve7o)*

**Comment by bileschi** [comment]
I think this is ready now.  Thanks for your patience.

**Comment by caisq** [comment]
:lgtm_strong:

Thanks for this!

---

Review status: 0 of 9 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[custom-layer/custom_layer.js, line 33 at r4](https://reviewable.io/reviews/tensorflow/tfjs-examples/61#-LBNJ8ZRCH2E-7QiAxgG:-LBNJ8ZSCOR4RqnUZcJ9:b-dqit3w) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/3ef2b980e87eedd55982764a65045d5504da72ca/custom-layer/custom_layer.js#L33)):*
> ```
>  * TODO(bileschi): File a github issue for the loading / saving of custom
>  * layers.
> ```

Can this be resolved before this PR is merged?

---

*[custom-layer/package.json, line 12 at r4](https://reviewable.io/reviews/tensorflow/tfjs-examples/61#-LBNJqq_8IGGWZjdl3rd:-LBNJqq_8IGGWZjdl3re:bs146o5) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/3ef2b980e87eedd55982764a65045d5504da72ca/custom-layer/package.json#L12)):*
> ```
> 0.9.0
> ```

Use the latest version: `0.10.3`.

---

*[custom-layer/README.md, line 7 at r4](https://reviewable.io/reviews/tensorflow/tfjs-examples/61#-LBNJeYC5iMMpQMGZWsR:-LBNJeYC5iMMpQMGZWsS:bg0k2e6) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/3ef2b980e87eedd55982764a65045d5504da72ca/custom-layer/README.md#L7)):*
> ```
> `computeOutputShape` overides of the `Layer`.
> ```

Optional: Break lines at 80 chars wherever possible to be consistent with Google markdown style guide.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-examples/61#-:-LBNKGvGDaqEGm6QFYtD:b-tn9zsu)*

**Comment by bileschi** [comment]
Review status: 0 of 9 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[custom-layer/custom_layer.js, line 33 at r4](https://reviewable.io/reviews/tensorflow/tfjs-examples/61#-LBNJ8ZRCH2E-7QiAxgG:-LBSF-AaA7cUBZ357bK2:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/3ef2b980e87eedd55982764a65045d5504da72ca/custom-layer/custom_layer.js#L33)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
>  * TODO(bileschi): File a github issue for the loading / saving of custom
>  * layers.
> ```

Can this be resolved before this PR is merged?
</blockquote></details>

Done.

---

*[custom-layer/package.json, line 12 at r4](https://reviewable.io/reviews/tensorflow/tfjs-examples/61#-LBNJqq_8IGGWZjdl3rd:-LBSF_6E5w9atHm3bL2V:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/3ef2b980e87eedd55982764a65045d5504da72ca/custom-layer/package.json#L12)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> 0.9.0
> ```

Use the latest version: `0.10.3`.
</blockquote></details>

Done.

---

*[custom-layer/README.md, line 7 at r4](https://reviewable.io/reviews/tensorflow/tfjs-examples/61#-LBNJeYC5iMMpQMGZWsR:-LBSF_bt7SUeOX7iKzJo:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/3ef2b980e87eedd55982764a65045d5504da72ca/custom-layer/README.md#L7)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> `computeOutputShape` overides of the `Layer`.
> ```

Optional: Break lines at 80 chars wherever possible to be consistent with Google markdown style guide.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-examples/61)*
